### PR TITLE
expose ngc i18n options

### PR DESCRIPTION
**Pull request by tdesmet**
This allows you to pass the i18n file, format and locale to ngc from the cli command line

### CODE REVIEW

**Comment by tdesmet** [comment]
I signed it!

**Comment by hansl** [comment]
@vicb could you confirm (we can talk about it tomorrow) that this change will work with the next iterations of i18n?

File: tests/e2e/tests/build/aot-i18n.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+               <source>Hello i18n!</source>
+               <target>Bonjour i18n!</target>
+               <note priority="1" from="description">An introduction header for this sample</note>
+               <note priority="1" from="meaning">User welcome</note>
+             </trans-unit>
+           </body>
+         </file>
+        </xliff>`))
+    .then(() => appendToFile('src/app/app.component.html',
+      '<h1 i18n="User welcome|An introduction header for this sample">Hello i18n!</h1>'))
```
Comment (vicb) [inline tests/e2e/tests/build/aot-i18n.ts]: remove `User welcome|` (because meaning should seldom be used)

**Comment by vicb** [comment]
yep let's talk tomorrow.

File: tests/e2e/tests/build/aot-i18n.ts (hunk 1 above)
Comment (tdesmet) [inline tests/e2e/tests/build/aot-i18n.ts]: Ok, I did not know that.

Also couldn't find anything about using it seldom on the angular cookbook.

File: tests/e2e/tests/build/aot-i18n.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+               <note priority="1" from="description">An introduction header for this sample</note>
+             </trans-unit>
+           </body>
+         </file>
+        </xliff>`))
+    .then(() => appendToFile('src/app/app.component.html',
+      '<h1 i18n="An introduction header for this sample">Hello i18n!</h1>'))
+    .then(() => ng('build', '--aot', '--i18n-file', 'src/locale/messages.fr.xlf', '--i18n-format',
+      'xlf', '--locale', 'fr'))
+    .then(() => expectFileToMatch('dist/main.bundle.js', /Bonjour i18n!/));
```
Comment (hansl) [inline tests/e2e/tests/build/aot-i18n.ts]: Could you add a negative test? Running `ng build` on the same project without `--locale` should not contain `Bonjour i18n`.

**Comment by tdesmet** [comment]
@hansl is this ok now?

**Comment by hansl** [comment]
I'll rebase and merge this soon.

**Comment by shaneog** [comment]
Just a little nudge. 
I have an application in production using CLI, AoT and i18n and at the moment I have to run a hacky gulp task to make it work. It would be great to get this merged.

Thanks!

**Comment by shaneog** [comment]
@hansl Is there anything we can do to get this PR into the next release?

**Comment by hansl** [comment]
Sorry, not beta 22, but beta 23. I'll rebase this today.

**Comment by rolandoldengarm** [comment]
Not sure if there's something wrong with the PR, but I've cloned the repo, linked it with "npm link", but when I run it, it is looking for a folder in your user directory @tdesmet  ?

````
$ ng serve --aot --locale zh --i18nFormat xtb --i18nFile ./src/i18n/messages.zh.xtb
** NG Live Development Server is running on http://localhost:4200. **
  0% compilingpath must be a string or Buffer
TypeError: path must be a string or Buffer
    at TypeError (native)
    at Object.fs.openSync (fs.js:640:18)
    at Object.fs.readFileSync (fs.js:508:33)
    at Function.CodeGenerator.create (C:\Users\iminar\Dev\angular\modules\@angular\compiler-cli\src\codegen.ts:102:29)
    at AotPlugin._make (C:\Projects\npm\angular-cli\packages\webpack\src\plugin.ts:207:54)
````
I have no clue why this is happening?

**Comment by tdesmet** [comment]
Have you tried putting the path between double quotes? 

`ng serve --aot --locale zh --i18nFormat xtb --i18nFile "./src/i18n/messages.zh.xtb"`

**Comment by rolandoldengarm** [comment]
Nope, same problem :( @tdesmet 

````
$ ng serve --aot --locale zh --i18nFormat xtb --i18nFile "./src/i18n/messages.zh.xtb"
** NG Live Development Server is running on http://localhost:4200. **
  0% compilingpath must be a string or Buffer
TypeError: path must be a string or Buffer
    at TypeError (native)
    at Object.fs.openSync (fs.js:640:18)
    at Object.fs.readFileSync (fs.js:508:33)
    at Function.CodeGenerator.create (C:\Users\iminar\Dev\angular\modules\@angular\compiler-cli\src\codegen.ts:102:29)
````

**Comment by filipesilva** [comment]
@rolandoldengarm that's probably a typescript problem, yesterday a few of those were cropping up. Make sure you are using `2.0.x` and NOT `2.1.x`.

**Comment by filipesilva** [comment]
@tdesmet some recent changes introduced a lot of conflicts with this PR, can you rebase and fix them?

**Comment by tdesmet** [comment]
Done

**Comment by rolandoldengarm** [comment]
@filipesilva I've tried on a clean machine, with TS 2.0.10, and still the same error.
I've cloned repo, used "npm link" to link it, then "npm link angular-cli" in my project.

**Comment by rolandoldengarm** [comment]
sorry for bothering you again, but when will this PR be merged and released? We are blocked by this; I don't know how to implement i18n + AOT without this PR being merged.

**Comment by tdesmet** [comment]
@hansl @filipesilva can you guys merge this please?

File: packages/angular-cli/commands/build.ts.orig [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+export interface BuildOptions {
+  target?: string;
+  environment?: string;
+  outputPath?: string;
+  watch?: boolean;
+  watcher?: string;
+  supressSizes: boolean;
+  baseHref?: string;
+  aot?: boolean;
+<<<<<<< HEAD
```
Comment (tdesmet) [inline packages/angular-cli/commands/build.ts.orig]: Woops, I'm sorry these .orig files should never have been committed. 

These are artifacts from merging and shouldn't have been added I'm used to them being ignored. 

Anyway I Removed them and everything should be good.

**Comment by ghidoz** [comment]
Hi guys, thanks for the good work! Do you know when it will be merged and released? Are there any chances for the next week? Thanks!

**Comment by hansl** [comment]
Alright. Merging this. I am about to change the plugin for the next release and this LGTM, so expect it to come in the next release (which will also be ng 2.3 compatible).

**Comment by hansl** [comment]
@tdesmet could we also have an angular-cli global option for these? would that be useful?

**Comment by tdesmet** [comment]
@hansl what do you mean?

**Comment by hansl** [comment]
Would it make sense to have `i18file` and `i18format` (and potentially `locale` as well?) in the `angular-cli.json` configuration for your app rather than having both on the command line?

Basically are those flags likely to change between two calls of `ng build`? I would think locale does, but i18file and format would not?

**Comment by tdesmet** [comment]
Format stays the same but locale and file do not in this implementation. File is translated messages for the locale. So if locale is fr then file is messages.fr.xlf anf if locale is de then file is messages.de.xlf.

I'm thinking we could probably make this easier by determining the file name by the given locale. Bu this would be less flexible as all users need to follow the same naming pattern.

Currently in my app I have a locales directory and each locale is a subdirectory containing a messages.locale.xlf file. My build process scans the locales directory and executes the build command for each found subdirectory. 
By typing this I realise I could just name all xlf files the same since they all live in seperate directories.

For me it would be great if we could implement something similar in the cli. But not sure if this workflow would work for others...

**Comment by hansl** [comment]
Having a CLI specific workflow is okay with me, as long as it’s easy and
transparent to the user. Let’s discuss more about this in a separate issue;
it’s probably not something coming to CLI final but I’m definitely
interested in making it easy for users to translate their apps in the
future! For now, having a broken workflow is probably good enough.

Thanks again for the PR!

-- 
Hans Larsen

On December 14, 2016 at 1:25:20 PM, Tom De Smet (notifications@github.com)
wrote:

Format stays the same but locale and file do not in this implementation.
File is translated messages for the locale. So if locale is fr then file is
messages.fr.xlf anf if locale is de then file is messages.de.xlf.

I'm thinking we could probably make this easier by determining the file
name by the given locale. Bu this would be less flexible as all users need
to follow the same naming pattern.

Currently in my app I have a locales directory and each locale is a
subdirectory containing a messages.locale.xlf file. My build process scans
the locales directory and executes the build command for each found
subdirectory.
By typing this I realise I could just name all xlf files the same since
they all live in seperate directories.

For me it would be great if we could implement something similar in the
cli. But not sure if this workflow would work for others...

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/angular/angular-cli/pull/3098#issuecomment-267161199>,
or mute the thread

**Pull request merged**